### ISSUE TITLE
Chrome timeline for Keras?

### ISSUE THREAD

**Issue by stalagmite7** [tensorflow/tensorflow#9868]
I've been looking at the chrome timeline to profile my Keras models, but have been unable to find any documentation on how I could use it with my Keras models.
I am running keras with a tf backend, and my sequential models are all built in keras. For instance, this is how my model is being built:

    'model.fit_generator( \
				generator= data_gen(args, 1), \
			 	steps_per_epoch=tr_steps, \
			 	epochs=args.epochs, \
			 	validation_data=data_gen(args, 2), \
			 	validation_steps=val_steps, \
			 	verbose=1, \
			 	callbacks=[checkpointer])`

 

I am at a loss for how I would try to generate a timeline trace for this model, and wanted to know if there is any related documentation for how I can do this?

**Comment by yaroslavvb** [comment]
This does not seem to be supported.

**Comment by yaroslavvb** [comment]
You could probably hack it by monkey patching 'tf.Session' to always create session with full statistics collection (as in https://github.com/tensorflow/tensorflow/issues/7794#issuecomment-281888201) and also monkey patching the "run" call to dump timeline file each time session.run gets called

**Comment by al626** [comment]
This was fixed in https://github.com/fchollet/keras/pull/6693

A simple example:

```python
from tensorflow.python.client import timeline

# Make your keras model
# ...
run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()
model.compile(loss='MSE', optimizer='Adam', options=run_options, run_metadata=run_metadata)
# Run model in your usual way
# ...
trace = timeline.Timeline(step_stats=run_metadata.step_stats)
with open('timeline.ctf.json', 'w') as f:
    f.write(trace.generate_chrome_trace_format())
```

**Comment by yaroslavvb** [comment]
@al626 training model issues multiple run calls, does this save metadata for last run call or does it merge them?

**Comment by al626** [comment]
Saves the last run call - I figured if people run multiple times for warm-up, then only the last run is relevant.

**Comment by Bowiemb** [comment]
@al626 What if you want the metadata for all run calls?

**Comment by soaxelbrooke** [comment]
After finding the commit that added the ability to trace, I came here looking for an example implementation - maybe a good candidate to go into the docs somewhere?

**Comment by mas-dse-greina** [comment]
Is the commit in the latest version of Keras? Or, do I need to build from source? 

As far as example implementation I am trying something like this, but would appreciate some feedback:

`import tensorflow as tf

	from keras import backend as K
	
	sess = tf.Session(config=tf.ConfigProto(
       intra_op_parallelism_threads=16, inter_op_parallelism_threads=2))

	run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
	run_metadata = tf.RunMetadata()

	K.set_session(sess)

	history = model.fit_generator(generator = training_generator,
                    steps_per_epoch = 24800//BATCH_SIZE,
                    validation_data = validation_generator,
                    validation_steps = 6200//BATCH_SIZE,
                    verbose=1, callbacks=[model_checkpoint])

	from tensorflow.python.client import timeline

	# Create the Timeline object, and write it to a json
	tl = timeline.Timeline(step_stats=run_metadata.step_stats)
	ctf = tl.generate_chrome_trace_format()
	with open('timeline.json', 'w') as f:
	    f.write(ctf)`

**Comment by mas-dse-greina** [comment]
Can this be done with inference too? I mean can you pass the run_metadata into the .predict()? Or when you load the model from a saved file?

**Comment by sakaia** [comment]
I am trying to modify imdb_fasttext.py as follows (4points added). 
But the chrome shows simple timeline for timeline.keras.json.
Is this correct usage for chrome timeline ?

```Python
'''This example demonstrates the use of fasttext for text classification

Based on Joulin et al's paper:

Bags of Tricks for Efficient Text Classification
https://arxiv.org/abs/1607.01759

Results on IMDB datasets with uni and bi-gram embeddings:
    Uni-gram: 0.8813 test accuracy after 5 epochs. 8s/epoch on i7 cpu.
    Bi-gram : 0.9056 test accuracy after 5 epochs. 2s/epoch on GTx 980M gpu.
'''

from __future__ import print_function
import numpy as np

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Embedding
from keras.layers import GlobalAveragePooling1D
from keras.datasets import imdb
# Add RunMetadata for timeline 1/4
import tensorflow as tf
from tensorflow.python.client import timeline

def create_ngram_set(input_list, ngram_value=2):
    """
    Extract a set of n-grams from a list of integers.

    >>> create_ngram_set([1, 4, 9, 4, 1, 4], ngram_value=2)
    {(4, 9), (4, 1), (1, 4), (9, 4)}

    >>> create_ngram_set([1, 4, 9, 4, 1, 4], ngram_value=3)
    [(1, 4, 9), (4, 9, 4), (9, 4, 1), (4, 1, 4)]
    """
    return set(zip(*[input_list[i:] for i in range(ngram_value)]))


def add_ngram(sequences, token_indice, ngram_range=2):
    """
    Augment the input list of list (sequences) by appending n-grams values.

    Example: adding bi-gram
    >>> sequences = [[1, 3, 4, 5], [1, 3, 7, 9, 2]]
    >>> token_indice = {(1, 3): 1337, (9, 2): 42, (4, 5): 2017}
    >>> add_ngram(sequences, token_indice, ngram_range=2)
    [[1, 3, 4, 5, 1337, 2017], [1, 3, 7, 9, 2, 1337, 42]]

    Example: adding tri-gram
    >>> sequences = [[1, 3, 4, 5], [1, 3, 7, 9, 2]]
    >>> token_indice = {(1, 3): 1337, (9, 2): 42, (4, 5): 2017, (7, 9, 2): 2018}
    >>> add_ngram(sequences, token_indice, ngram_range=3)
    [[1, 3, 4, 5, 1337], [1, 3, 7, 9, 2, 1337, 2018]]
    """
    new_sequences = []
    for input_list in sequences:
        new_list = input_list[:]
        for i in range(len(new_list) - ngram_range + 1):
            for ngram_value in range(2, ngram_range + 1):
                ngram = tuple(new_list[i:i + ngram_value])
                if ngram in token_indice:
                    new_list.append(token_indice[ngram])
        new_sequences.append(new_list)

    return new_sequences

# Set parameters:
# ngram_range = 2 will add bi-grams features
ngram_range = 1
max_features = 20000
maxlen = 400
batch_size = 32
embedding_dims = 50
epochs = 5

print('Loading data...')
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=max_features)
print(len(x_train), 'train sequences')
print(len(x_test), 'test sequences')
print('Average train sequence length: {}'.format(np.mean(list(map(len, x_train)), dtype=int)))
print('Average test sequence length: {}'.format(np.mean(list(map(len, x_test)), dtype=int)))

if ngram_range > 1:
    print('Adding {}-gram features'.format(ngram_range))
    # Create set of unique n-gram from the training set.
    ngram_set = set()
    for input_list in x_train:
        for i in range(2, ngram_range + 1):
            set_of_ngram = create_ngram_set(input_list, ngram_value=i)
            ngram_set.update(set_of_ngram)

    # Dictionary mapping n-gram token to a unique integer.
    # Integer values are greater than max_features in order
    # to avoid collision with existing features.
    start_index = max_features + 1
    token_indice = {v: k + start_index for k, v in enumerate(ngram_set)}
    indice_token = {token_indice[k]: k for k in token_indice}

    # max_features is the highest integer that could be found in the dataset.
    max_features = np.max(list(indice_token.keys())) + 1

    # Augmenting x_train and x_test with n-grams features
    x_train = add_ngram(x_train, token_indice, ngram_range)
    x_test = add_ngram(x_test, token_indice, ngram_range)
    print('Average train sequence length: {}'.format(np.mean(list(map(len, x_train)), dtype=int)))
    print('Average test sequence length: {}'.format(np.mean(list(map(len, x_test)), dtype=int)))

print('Pad sequences (samples x time)')
x_train = sequence.pad_sequences(x_train, maxlen=maxlen)
x_test = sequence.pad_sequences(x_test, maxlen=maxlen)
print('x_train shape:', x_train.shape)
print('x_test shape:', x_test.shape)

print('Build model...')
model = Sequential()

# we start off with an efficient embedding layer which maps
# our vocab indices into embedding_dims dimensions
model.add(Embedding(max_features,
                    embedding_dims,
                    input_length=maxlen))

# we add a GlobalAveragePooling1D, which will average the embeddings
# of all words in the document
model.add(GlobalAveragePooling1D())

# We project onto a single unit output layer, and squash it with a sigmoid:
model.add(Dense(1, activation='sigmoid'))

# Add RunMetadata for timeline 2/4
run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'],
# Add RunMetadata for timeline 3/4
              options=run_options,
              run_metadata=run_metadata)

model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(x_test, y_test))

# Add RunMetadata for timeline 4/4
trace = timeline.Timeline(step_stats=run_metadata.step_stats)
with open('timeline.keras.json', 'w') as f:
    f.write(trace.generate_chrome_trace_format())
```

**Comment by JaspervDalen** [comment]
@al626 I try to get a timeline but it creates a file that only contains:
`    "traceEvents": [
        {
            "name": "process_name",
            "ph": "M",
            "pid": 0,
            "args": {
                "name": "Allocators"
            }
        }
    ]
}'

I git this by running the following
`
    "traceEvents": [
        {
            "name": "process_name",
            "ph": "M",
            "pid": 0,
            "args": {
                "name": "Allocators"
            }
        }
    ]
}
`

**Comment by lmxhappy** [comment]
just after compile, run timeline.Timeline(step_stats=run_metadata.step_stats) . is ok?@al626

**Comment by kenfehling** [comment]
tensorflow.keras.Model doesn't allow run_metadata correct?

**Comment by nicoulaj** [comment]
I may be wrong but it seems this was broken, at least since commit 24e7ac0756aeba06ce1cf7f7e72219a11b14b101, which adds this check here: https://github.com/keras-team/keras/commit/24e7ac0756aeba06ce1cf7f7e72219a11b14b101#diff-a18b8c6a1191d6f49303e0d599ca8c37R2485

**Comment by jonathanking** [comment]
@kenfehling You are correct, see pull request #19932

**Comment by psteinb** [comment]
recording timelines appears to work with tf1.11 and keras 2.2.4. I agree to @bhack that some more docs are needed. 

But the timeline appears surprisingly short. I did a simple CNN with 12 epochs which takes about 1.5 minutes to complete on a Titan Xp. But the timeline only shows a record up to 3.5s (the number of convolutions is also a bit off, 12 epochs on 60k samples with a batch size of 32 and I only see 14 Conv2D calls - is tf optimizing the compute graph that badly?). 

I put the timeline and the keras script that produced it into this [gist](https://gist.github.com/psteinb/d3315a065a90814b435ac024e9835aab) so that people can reproduce if needed.

**Comment by psteinb** [comment]
here is a screen shot of a `nvprof` recorded run of the same script. One can clearly see the 12 epochs, so I am inclined to say that the timeline is incomplete.
![nvvp-screenshot_2018-10-09_14-11-39](https://user-images.githubusercontent.com/1465603/46668725-9bf41900-cbcd-11e8-9726-352a57450e7c.png)
It's also interesting that tf decided to use `maxwell_winograd` style convolutions on a Titan Xp which is a Pascal card. I need to dig into this.

**Comment by rohith14** [comment]
@psteinb :  Do you more info on why maxwell_scudnn style convolutions on Titan Xp?  I am also seeing the same for a different use-case.  Thanks

**Comment by psteinb** [comment]
@rohith14 : what do you mean? Do I have more info on this or do I want more details on this? I want more information on this, if anyone could share them.

**Comment by rohith14** [comment]
@psteinb I am looking for more information on this.  I am seeing maxwell_scudnn style convolutions on Titan Xp and not sure why.

**Comment by psteinb** [comment]
@rohith14 What cuda and tf version did you use?

**Comment by Barbarioli** [comment]
@mas-dse-greina have you found a way to use it during inference? Is it possible to do it?

**Comment by tonyreina** [comment]
https://gist.github.com/tonyreina/5bbe050c2cfceae62a1dda7d9010b692

Yeah. I ended up creating this which just runs the Keras model in a regular TensorFlow session.  It seems to work.

**Comment by mohantym** [comment]
Hi @stalagmite7 ! 
We are checking to see whether you still need help in this issue .Have you checked above[ comment](https://github.com/tensorflow/tensorflow/issues/9868#issuecomment-499695742) yet?
